packages/checkbox/src/index.js

```diff
@@ -0,0 +1,266 @@
+////////////////////////////////////////////////////////////////////////////////
+// Welcome to @reach/checkbox!
+
+import React, {
+  forwardRef,
+  createContext,
+  useContext,
+  useEffect,
+  useRef,
+  useState
+} from "react";
+import { wrapEvent, assignRef } from "@reach/utils";
+import warning from "warning";
+import { any, string, bool, func, oneOfType, node, shape } from "prop-types";
+
+const CustomCheckboxContext = createContext({});
+const useCheckboxContext = () => useContext(CustomCheckboxContext);
+
+////////////////////////////////////////////////////////////////////////////////
+export const CustomCheckboxContainer = forwardRef(
+  function CustomCheckboxContainer({ children, ...props }, forwardedRef) {
+    const [
+      { checked, mixed, disabled, readOnly, focused },
+      setContainerState
+    ] = useState({
+      checked: null,
+      mixed: null,
+      disabled: null,
+      readOnly: null,
+      focused: null
+    });
+
+    return (
+      <CustomCheckboxContext.Provider
+        value={{ setContainerState, visuallyHidden: true }}
+      >
+        <div
+          ref={forwardedRef}
+          {...props}
+          data-reach-custom-checkbox-container=""
+          data-checked={checked ? "" : undefined}
+          data-focused={focused ? "" : undefined}
+          data-mixed={mixed ? "" : undefined}
+          data-disabled={disabled ? "" : undefined}
+          data-read-only={readOnly ? "" : undefined}
+        >
+          {children}
+        </div>
+      </CustomCheckboxContext.Provider>
+    );
+  }
+);
+
+////////////////////////////////////////////////////////////////////////////////
+export const CustomCheckbox = forwardRef(function CustomCheckbox(
+  {
+    checked: controlledChecked,
+    defaultChecked,
+    disabled: disabledProp,
+    checkmarks,
+    children,
+    id,
+    name,
+    onChange,
+    readOnly,
+    value,
+    ...props
+  },
+  forwardedRef
+) {
+  const [inputProps, { checked }] = useMixedCheckbox({
+    checked: controlledChecked,
+    defaultChecked,
+    disabled: disabledProp,
+    onChange,
+    readOnly
+  });
+  return (
+    <CustomCheckboxContainer
+      data-reach-custom-checkbox=""
+      {...props}
+      ref={forwardedRef}
+    >
+      {checked === true && checkmarks.true}
+      {checked === false && checkmarks.false}
+      {checked === "mixed" && checkmarks.mixed}
+      {children}
+      <MixedCheckbox
+        id={id}
+        value={value}
+        name={name}
+        readOnly={readOnly}
+        {...inputProps}
+      />
+    </CustomCheckboxContainer>
+  );
+});
+
+if (__DEV__) {
+  CustomCheckbox.propTypes = {
+    checkmarks: shape({
+      true: node,
+      false: node,
+      mixed: node
+    })
```

Comment:
shouldn't this be required?

Reply:
I don't think so. You could style a custom checkbox without passing any additional components (with pseudo elements, for example).